### ISSUE TITLE
[Cppcheck] Support "Bug hunting"

### ISSUE THREAD

**Issue by ybiquitous** [sider/runners#1138]
> ### Bug hunting
>
> There is a new "soundy" analysis in Cppcheck that should detect most bugs. You should expect false alarms, however the false alarms will not be overwhelming.
>
> This new "soundy" analysis is not intended to replace normal Cppcheck analysis. > There are use cases where false alarms can not be tolerated.
>
> We have added 1 checker and that checks for division by zero:
>
>    It detects all "integer division by zero" bugs in the Juliet test suite.
>    It detects all "division by zero" bugs in the ITC test suite.
>    There was 28 division by zero CVEs published in 2019 for C/C++ open source projects, and we could quickly see that 21 of the bugs are found by Cppcheck. > There is no CVE bug that we know Cppcheck fails to diagnose. But there are 7 CVEs that would require additional investigation to establish if it is really detected or not.

https://github.com/danmar/cppcheck/releases/tag/2.0

New feature! 🎉 Let's support it!

See also:
https://github.com/danmar/cppcheck/blob/master/man/manual.md#bug-hunting

_Originally posted by @ybiquitous in https://github.com/sider/runners/pull/1083#issuecomment-629028050_

**Comment by ybiquitous** [comment]
This feature **cannot** be realized by just adding the `bug-hunting` option.
See the following reproduction:

```console
$ bundle exec rake docker:build docker:shell ANALYZER=cppcheck
$ cd /work/test/smokes/cppcheck/default
$ cppcheck --bug-hunting .
Checking arrayIndexOutOfBounds.c ...
1/5 files checked 14% done
Checking autoVariables.c ...
2/5 files checked 30% done
Checking good.c ...
3/5 files checked 35% done
Checking src/erase.cpp ...
4/5 files checked 87% done
Checking src/nullpointer.c ...
5/5 files checked 100% done
```

As you see the output of `cppcheck --bug-hunting .` command, normal issues are not reported with `--bug-hunting`.

**Comment by suetsuguk** [comment]
As for Z3, we need to install Z3 as commented at https://github.com/danmar/cppcheck/blob/master/readme.txt#L25-L31
However, we are able to compile it without Z3. Both results for the sample CVEs are identical.

Note that the libz3-dev package on ubuntu:xenial-20190222(Debian Buster) is too old to compile cppcheck V2.0.
Currently we need to get the latest release archive from https://github.com/Z3Prover/z3/releases

* with Z3
$ make clean
$ make USE_Z3=yes LDFLAGS=-L/home/foo/z3-4.8.8-x64-ubuntu-16.04/bin/ CPPFLAGS=-I/home/foo/z3-4.8.8-x64-ubuntu-16.04/include/
$ ../../../cppcheck --bug-hunting . > result_z3.txt 2>&1
$ wc result_z3.txt
  2753  17967 181262 result_z3.txt
$ ldd cppcheck
        linux-vdso.so.1 (0x00007ffe18d85000)
        libz3.so => /home/foo/z3-4.8.8-x64-ubuntu-16.04/bin/libz3.so (0x00007f1b553ea000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f1b5525b000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1b550d8000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f1b550be000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1b54efd000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1b54edc000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1b57769000)

* without Z3
$ make clean
$ make
$ ../../../cppcheck --bug-hunting . > result_wo_z3.txt 2>&1
$ wc result_wo_z3.txt
  2753  17967 181262 result_wo_z3.txt
$ ldd cppcheck
        linux-vdso.so.1 (0x00007ffca05a4000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0a654e5000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0a65362000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f0a65348000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0a65187000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0a6622b000)

$ diff result_z3.txt result_wo_z3.txt
(no difference)

**Comment by suetsuguk** [comment]
As for the `--bug-hunting` flag, cppcheck executes the only Bug Hunting check if it is specified as seen at https://github.com/danmar/cppcheck/blob/master/lib/cppcheck.cpp#L938-L942
A simple solution for us to support both the normal check and the Bug Hunting is to run `cppcheck` twice and combine the results.

**Comment by ybiquitous** [comment]
Thank you so much for the detailed explanation.

> However, we are able to compile it without Z3. Both results for the sample CVEs are identical.

It seems better to me that we follow the Cppcheck's instruction to compile with Z3.

> A simple solution for us to support both the normal check and the Bug Hunting is to run `cppcheck` twice and combine the results.

I agree with this idea.

**Comment by suetsuguk** [comment]
I might have made a mistake.

1. CPPFLAGS
Passing the CPPFLAGS value to the `make` command as a command-line argument, the CPPFLAGS value in `Makefile` is overwritten by the argument. So the `USE_Z3=yes` flag does not work. That is why the results are identical.

2. cppcheck Version 2.0
Compiling cppcheck Version 2.0 with the latest version of z3-4.8.8, we need to define `-DNEW_Z3` flag as commented at https://sourceforge.net/p/cppcheck/discussion/development/thread/e412731255/?limit=100

3. segmentation fault
Finally I was able to compile cppcheck v2.0 with z3 v.4.8.8. I tried to run `--bug-hunting` but failed with segmentation fault.
I also found cppcheck compiled with the debian package of `libz3-dev (v.4.4.1)` worked fine. It reported more issues than "w/o z3".

Here is the proper step to compile:
- install the debian package as commented at https://github.com/danmar/cppcheck/blob/master/readme.txt#L25-L26
- invoke `make MATCHCOMPILER=yes HAVE_RULES=yes USE_Z3=yes CPPFLAGS="-DUSE_Z3 -Wno-unused-function -Wno-deprecated-declarations"`
